캐시서버를 위한 Cache-Control 헤더 제어

wkpark opened the pull request:
- 캐시 친화적인 `Cache-Control` 헤더
- 로그인사용자는 `cache-control: private`, 그 이외에는 `public`을 사용
- `session_start()`를 로그인 할 경우에만 사용하여 `set-cookie` 사용 최소화

모니위키에서 varnish 캐시서버 지원을 넣었던 경험을 바탕으로 일단 XE에서 최소한의 변경을 가해봤습니다.

모니위키의 경우에 vanish 캐시서버 - 아파치 상태에서 `ab -n 1000 -c 3`을 평균적인 크기의 페이지에 대해 테스트해보면 15000~20000 RPS나 나옵니다. (varnish 캐시 없이는 150~300 RPS)

반면 XE의 경우에는 `Set-Cookie`등의 요인을 제거하면 (이 패치 및 set-cookie 패치)
`ngnix` 웹서버상에서 80 RPS 정도가 나오며, varnish 캐시서버를 쓰면 5000RPS~15000RPS가 나옵니다.
# nginx

```
$ ab -n 500 -c 3 http://localhost:8000/ (nginx)
Requests per second:    82.90 [#/sec] (mean)
Time per request:       36.187 [ms] (mean)
Time per request:       12.062 [ms] (mean, across all concurrent requests)
Transfer rate:          2706.19 [Kbytes/sec] received
...
```
# varnish + nginx

```
$ ab -n 500 -c 3 http://localhost/
...
Requests per second:    6797.36 [#/sec] (mean)
Time per request:       0.441 [ms] (mean)
Time per request:       0.147 [ms] (mean, across all concurrent requests)
Transfer rate:          222706.38 [Kbytes/sec] received
...

$ ab -n 5000 -c 100 http://localhost/
Requests per second:    13301.13 [#/sec] (mean)
Time per request:       7.518 [ms] (mean)
Time per request:       0.075 [ms] (mean, across all concurrent requests)
Transfer rate:          435806.75 [Kbytes/sec] received
...
```

wkpark committed:
Context::setCacheControl() method added etc.

 * make cache friendly
 * make cache-control public/private conditionally
 * use session_start() conditionally

wkpark commented:
이 패치는 XE의 구조를 정확히 모르고 급조한 것이므로 이를 참조하여 개선하셨으면 합니다~

wkpark commented:
#1596 참조

misol commented:
세션을 로그인에만 사용하는 것은 아닌데.. session_start를 doLogin 에만 사용하는 것은 무리수 아닌지요..

wkpark commented:
`session_start()`가 호출 되어야 세션 쿠기 `$_SESSION`을 비로소 사용할 수 있게 됩니다.
따라서 `$_SESSION` 세션쿠키가 어디서 사용되는지 추적하면 답이 나오겠죠. `$_COOKE`는 세션과 상관 없습니다.

그리고 비로그인 사용자의 경우라면 `$_SESSION` 세션쿠키를 쓸것이 아니라 보안에 덜 민감한 `$_COOKIE`를 쓰도록 고치는 것이 마땅해보입니다.

또한, 요즘 배포판은 `session.use_cookies=1` 및  `session.use_only_cookies = 1` 이 기본으로 설정이 되어있습니다. 세션 ID 전달을 쿠키를 통해서만 하고 `GET` / `POST` 인자로 전달하지 않는 것이지요. 쿠키라고 해서 딱히 보안에 나쁜 것이 아닌 것이며, 어떤 식으로 쿠키를 사용하는지 더 중요한 것입니다.

wkpark commented:
세션이 사용되는 곳은 34개 파일, 228 줄이며, 세션쿠키 이름은 47개로 다음과 같습니다

`$_SESSION['INPUT_ERROR']`
`$_SESSION['XE_VALIDATOR_ERROR']`
`$_SESSION['XE_VALIDATOR_ERROR_LANG']`
`$_SESSION['XE_VALIDATOR_ID']`
`$_SESSION['XE_VALIDATOR_MESSAGE']`
`$_SESSION['XE_VALIDATOR_MESSAGE_TYPE']`
`$_SESSION['XE_VALIDATOR_RETURN_URL']`
`$_SESSION['__XE_EASYINSTALL_REDIRECT__']`
`$_SESSION['__XE_FILE_KEY_AND__']`
`$_SESSION['__XE_FILE_KEY__']`
`$_SESSION['__XE_UPLOADING_FILES_INFO__']`
`$_SESSION['_editor_sequence_']`
`$_SESSION['accessible']`
`$_SESSION['accessibled_comment']`
`$_SESSION['auth_member_info']`
`$_SESSION['auth_member_srl']`
`$_SESSION['avoid_log']`
`$_SESSION['banned_document']`
`$_SESSION['captcha_authed']`
`$_SESSION['captcha_keyword']`
`$_SESSION['comment_management']`
`$_SESSION['declared_comment']`
`$_SESSION['declared_document']`
`$_SESSION['denied_admin']`
`$_SESSION['document_management']`
`$_SESSION['enviroment_gather']`
`$_SESSION['file_management']`
`$_SESSION['ftp_password']`
`$_SESSION['group_srls']`
`$_SESSION['ipaddress']`
`$_SESSION['is_admin']`
`$_SESSION['is_logged']`
`$_SESSION['lang_type']`
`$_SESSION['member_captcha_authed']`
`$_SESSION['member_srl']`
`$_SESSION['own_comment']`
`$_SESSION['own_document']`
`$_SESSION['poll_management']`
`$_SESSION['readed_document']`
`$_SESSION['rechecked_password_step']`
`$_SESSION['seq']`
`$_SESSION['upload_info']`
`$_SESSION['use_rewrite']`
`$_SESSION['voted_comment']`
`$_SESSION['voted_document']`
`$_SESSION['xe_logged']`
`$_SESSION['xe_temp_password_' . $user_id]`

이중에 상당수는 로그인 사용자 전용이죠.

misol commented:
"상당수" 일뿐 전부는 아니기 때문에 문제가 생길 가능성이 있습니다.

서드파티 프로그램들 까지 고려하면 가능성이 더 큽니다.

당장 캡챠 애드온만 해도 문제가 됩니다.

캡챠를 쿠키로 운영하면 문제가 생기기 때문에 쿠키로 운영할 수도 없습니다.




On Tue, Jul 7, 2015 at 10:14 PM -0700, "Won Kyu Park" <notifications@github.com> wrote:












세션이 사용되는 곳은 34개 파일, 228 줄이며, 세션쿠키 이름은 47개로 다음과 같습니다



$_SESSION['INPUT_ERROR']

$_SESSION['XE_VALIDATOR_ERROR']

$_SESSION['XE_VALIDATOR_ERROR_LANG']

$_SESSION['XE_VALIDATOR_ID']

$_SESSION['XE_VALIDATOR_MESSAGE']

$_SESSION['XE_VALIDATOR_MESSAGE_TYPE']

$_SESSION['XE_VALIDATOR_RETURN_URL']

$_SESSION['__XE_EASYINSTALL_REDIRECT__']

$_SESSION['__XE_FILE_KEY_AND__']

$_SESSION['__XE_FILE_KEY__']

$_SESSION['__XE_UPLOADING_FILES_INFO__']

$_SESSION['_editor_sequence_']

$_SESSION['accessible']

$_SESSION['accessibled_comment']

$_SESSION['auth_member_info']

$_SESSION['auth_member_srl']

$_SESSION['avoid_log']

$_SESSION['banned_document']

$_SESSION['captcha_authed']

$_SESSION['captcha_keyword']

$_SESSION['comment_management']

$_SESSION['declared_comment']

$_SESSION['declared_document']

$_SESSION['denied_admin']

$_SESSION['document_management']

$_SESSION['enviroment_gather']

$_SESSION['file_management']

$_SESSION['ftp_password']

$_SESSION['group_srls']

$_SESSION['ipaddress']

$_SESSION['is_admin']

$_SESSION['is_logged']

$_SESSION['lang_type']

$_SESSION['member_captcha_authed']

$_SESSION['member_srl']

$_SESSION['own_comment']

$_SESSION['own_document']

$_SESSION['poll_management']

$_SESSION['readed_document']

$_SESSION['rechecked_password_step']

$_SESSION['seq']

$_SESSION['upload_info']

$_SESSION['use_rewrite']

$_SESSION['voted_comment']

$_SESSION['voted_document']

$_SESSION['xe_logged']

$_SESSION['xe_temp_password_' . $user_id]



이중에 상당수는 로그인 사용자 전용이죠.



—
Reply to this email directly or view it on GitHub.

wkpark commented:
세션 / 쿠키 관련된 수정은 Varnish 캐시서버를 제대로 적용하기 위한 가장 기본 단계입니다. 이걸 해결해도 완벽한 varnish 캐시 서버를지원하기에 할일이 많을 것입니다만, 이정도만 수정하면 말그대로 빛의 속도 XE가 될 수 있습니다.

wkpark commented:
> "상당수" 일뿐 전부는 아니기 때문에 문제가 생길 가능성이 있습니다.
>
> 서드파티 프로그램들 까지 고려하면 가능성이 더 큽니다.
> 당장 캡챠 애드온만 해도 문제가 됩니다.
> 캡챠를 쿠키로 운영하면 문제가 생기기 때문에 쿠키로 운영할 수도 없습니다.

그러니 더욱 더 빨리 수정하고, 가이드를 제시해야겠지요.

캅챠 애드온 / 파일 모듈 / (자동)설치모듈 / 스팸필터 / 커멘트 모듈 이정도를 수정하거나 확인이 필요하겠네요.

YJSoft commented:
@wkpark 얼마나 많은 서드파티 자료가 비로그인 상태에서 세션을 사용하는지 알 수 없습니다.
물론 state값 등 보안에 크게 중요하지 않다면 굳이 세션을 쓸 것이 아니라 쿠키를 사용해도 되겠지만, 호환성이 문제입니다. 예전 XE 1.4->1.5때보다도 훨씬 큰 문제가 발생할 수 있습니다.

wkpark commented:
일단 옵션을 두어서 `$db_info->use_session_conditionally`가 정의되었을 경우에만 `session_start()`를 호출하지 않도록 하였습니다. 따라서 이 패치를 적용하더라도 실제로 `$db_info->use_session_conditionally`를 정의하지 않으면 예전과 `session_start()`가 항상 호출됩니다.

세션관련된 수정을 하루아침에 고칠 수는 없는 노릇이나, varnish 캐시서버의 빛과같은 속도가 중요한 경우에는 우선 이 옵션을 쓰면 될것이요, 세션을 쓰는 중요한 모듈이 필수적인 경우에는 이를 안쓰도록 하면 될것입니다.

또한 `session_start()`를 아예 안쓰자는 것이 아닙니다. 세션 관련된 문제가 캐시서버를 방해하니 줄이자는 것이고, 또한 현재 무분별하게 쓰이는 세션을 최대한 줄일 수 있다는 것입니다.

wkpark commented:
세션에 관련된 모듈에 대하여 세션을 사용중인지 검사하고 `session_start()`를 호출하게끔 고쳤습니다. `$db_info->use_session_conditionally=true;`같은 식으로 환경변수를 정의하고, 파일 업로드 / 캅챠 등등이 작동하는 것을 확인하였습니다.

확인하는 방법은 간단합니다. 쿠키에서 `PHPSESSID`를 지웁니다. 그런 다음에 원하는 페이지 (사용하는 애드온을 쓰고 있는)를 가면 `session_start()`가 정상 호출되면 `PHPSESSID` 쿠키가 자동 세팅됩니다.

`PHPSESSID` 쿠키가 자동으로 세팅이 안된다면 `session_start()`가 호출 안된 것이고, 이를 적당한 곳에 (`init()` 등등에서 호출하게 만들면 됩니다.

wkpark committed:
do not always set mobile/user-agent cookies

qw5414 commented:
https://github.com/xpressengine/xe-core/blob/master/CONTRIBUTING.md

전체적인 코드가 XE에서 사용하는 코딩콘벤션에 어긋나는것 같아요.. 한번 확인해보시고, 그대로 적용해주시는것이 좋을 것 같아요..

그리고, 저도 세션이 마구잡이로 다시 시작되는건.. 좋은 생각은 아닌것 같아요. 세션정보를 이용한 서드파티가 꾀 되고.. DB세션도 한번 테스트 해볼 필요가 있거든요..

기존에 XE에서 세션에 대한 코드 적용된 부분도 현재는 적용을 재대로 하지 못햇던 이유가 여러 서드파티들의 호환성때문이기도 했고요..

wkpark commented:
@qw5414 // 네 코딩컨벤션을 또 안지켰네요 ^^;;
28017b774c 커밋에서 말하고자 한 것은 `session_start()`의 무조건적인 호출을 배제하고, 모듈별로 선별적으로 한줄의 코드를 추가하여 해결 가능하다는 것을 보여주는 목적으로 올린 것입니다.

이 글 처음에도 밝혔듯이, 이 패치는 제가 XE 코드를 제대로 이해하고 작성한 것이 아니므로, 이러한 변경을 가하면 이러이러한 성능의 향상이 있다는 것을 보여주는 예시라고 생각하시면 되겠습니다.

이를 바탕으로 좀 더 좋은 코드를 만들어주시면 좋겠습니다.
- - - -
기타 Varnish 캐시 서버에 친화적인 캐싱을 고려하려면 다음을 고려해야 할것입니다.

컨셉은 단순합니다. 각 페이지를 static하게 만드는 것입니다. 그렇게 하려면 동적인 컨텐츠 부분은 javascript로 넘기고, 일부 빈번하게 업데이트 되는 내용은 ajax 혹은 자바스크립트를 include하는 형태로 분리해서 처리하는 것입니다. ESI처리는 범용성이 떨어지니 논외로 합니다. 아예 조회수같은 것은 빼버리는 방법도 있을 것입니다. 혹은 조회수가 들어가면 `Pragma:no-cache`를 넣어버리면 됩니다.

1. `ETag` 혹은 `Last-Modified` 헤더 설정: 게시글에 대해서는 게시글이 갱신된 시간에 대한 정의 즉, 커멘트가 최종적으로 달린 시간 / vote up/down이 갱신된 마지막 시간 / 글의 조회수를 별도로 ajax로 처리함.
2. 게시글 목록 역시 최종 글이 올라간 시간을 기준으로 하되, 조회수나 기타 정보는 별도의 ajax 처리.
3. 동적인 컨텐츠는 `Pragma:no-cache` 혹은 이에 준하는 캐시컨트롤 헤더 세팅

wkpark commented:
* `session_start()`가 실행이 되어야 `PHPSESSID`가 세팅이 된다. (즉 `Set-Cookie:` 헤더가 붙음)
* 혹은 `session_id($blahblah);`가 실행되면 강제로 `PHPSESSID`가 세팅이 된다.
* `PHPSESSID`쿠키가 없다 -> `session_start()`가 Context에서 실행되지 않음 -> `PHPSESSID` 세팅 안됨
* 로그인을 하거나 `session_start()`가 호출되는 경우 -> `PHPSESSID`가 세팅이 됨
* 한번 `PHPSESSID`가 세팅이 되면 -> `session_start()`가 매번 실행됨.-> 세션이 만료될때까지.

이러면 되는 시나리오이기때문에 매번 `session_start()`를 실행시킬 이유가 전혀 없으며,
또한 각 모듈별로 세션이 필요한 경우에는 `session_start()`를 유동적으로 호출시키면 되고 (한줄),
당분간은 `$db_info->use_session_conditionally` 옵션을 두어서 호환성 문제에 대한 유예기간을 두면서
`session_start()`관련 문제를 지속적으로 고쳐 나갈 수 있을 것입니다.

misol commented:
왜 세션을 로그인에만 사용한다고 생각하시죠?




On Wed, Jul 8, 2015 at 3:19 PM -0700, "Won Kyu Park" <notifications@github.com> wrote:












session_start()가 실행이 되어야 PHPSESSID가 세팅이 된다. (즉 Set-Cookie: 헤더가 붙음)
혹은 session_id($blahblah);가 실행되면 강제로 PHPSESSID가 세팅이 된다.

PHPSESSID쿠키가 없다 -> session_start()가 Context에서 실행되지 않음 -> PHPSESSID 세팅 안됨
로그인을 하거나 session_start()가 호출되는 경우 -> PHPSESSID가 세팅이 됨
한번 PHPSESSID가 세팅이 되면 -> session_start()가 매번 실행됨.-> 세션이 만료될때까지.




이러면 되는 시나리오이기때문에 매번 session_start()를 실행시킬 이유가 전혀 없으며,

또한 각 모듈별로 세션이 필요한 경우에는 session_start()를 유동적으로 호출시키면 되고 (한줄),

당분간은 $db_info->use_session_conditionally 옵션을 두어서 호환성 문제에 대한 유예기간을 두면서

session_start()관련 문제를 지속적으로 고쳐 나갈 수 있을 것입니다.



—
Reply to this email directly or view it on GitHub.

wkpark commented:
@misol // 위에 제 글 및 패치를 보셨는지요? 모듈별로 세션 사용할때 `session_start()` 한 줄 추가 하면 됩니다.

qw5414 commented:
@wkpark 모듈 서드파티 개발자가 이 패치의 영향으로 `session_start()` 를 추가하는 패치 해야한다는건가요?

그럼.. 수많은 모듈의 서드파티 개발자가 지원중단한 것도 있을테니. 이것이 패치된다면, 업데이트를 안하는게 좋겠군요. 모듈의 적용여부를 떠나서요..

wkpark commented:
@qw5414 // 장기적으로 봤을때에 `session_start()`를 선별적으로 호출하게 해야 한다는 것이고, 제가 위에 써놓은 내용를 보시면 `$db_info->use_session_conditionally` 옵션이 기본적으로 꺼져있는 상태이기때문에 이 패치를 적용하셔도 `session_start()`가 무조건 호출됩니다. 따라서 이 패치를 적용해도 일반 XE 운영자들은 이패치에 대해서 신경쓸 것이 전혀 없습니다.

물론 varnish 캐시 서버를 사용하려는 XE 운영자들은 당분간 모듈의 `session` 사용 여부를 보고, 고쳐야 할 경우에 고쳐야 하겠구요. `session_start()` 한줄 추가해 넣는 것이 어려운 것도 아닐분더라, 위의 설명에도 이마 나와있지만 그것을 `curl -I` 명령을 통해서 아주 손쉽게 파악하고 고치는 것이 가능합니다.

wkpark commented:
추가 varnish 캐시  테스트 정보입니다
~~~
$ tail -1 files/config/db.config.php
#$db_info->use_session_conditionally=true;
~~~
설정이 꺼져있습니다. (커멘트아웃 되어잇음) 즉 `session_start()`가 무조건적으로 호출됩니다. 따라서 varnish 캐시서버가 효율적으로 작동하지 않게 됩니다.
~~~
$ curl -I  http://localhost/
HTTP/1.1 200 OK
Content-Type: text/html; charset=UTF-8
X-Powered-By: PHP/5.5.9-1ubuntu4.9
Cache-Control: public, must-revalidate, post-check=0, pre-check=0
Set-Cookie: PHPSESSID=1g12hp22q8r0ojbjog7seaft26; path=/
Accept-Ranges: bytes
Date: Thu, 09 Jul 2015 00:54:35 GMT
X-Varnish: 1471733958
Age: 0
Via: 1.1 varnish
Connection: keep-alive
X-Backend-Server: nginx/1.4.x (Ubuntu)
Server: nginx/1.4.x (Ubuntu)
X-Cache-Hit: MISS
~~~
캐시 미스가 되고 있습니다.

ab 벤치마크입니다. 1000회를, 동시에 3개 요청하는 경우입니다. 상당히 느리고 이것이 현재 XE를 nginx + varnish 캐시서버로 운용할때의 대략적 성능입니다. (혹은 단순히 nginx만 운용하는 경우)
~~~~
$ ab -n 1000 -c 3  http://localhost/
This is ApacheBench, Version 2.3 <$Revision: 1528965 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking localhost (be patient)
Completed 100 requests
Completed 200 requests
Completed 300 requests
Completed 400 requests
....
Requests per second:    74.85 [#/sec] (mean)
Time per request:       40.079 [ms] (mean)
Time per request:       13.360 [ms] (mean, across all concurrent requests)
Transfer rate:          2533.88 [Kbytes/sec] received
...
~~~~
그러면 이번에는 캐시 설정을 켜보겠습니다. 즉 `session_start()`가 기본적으로 호출이 안되고 `Set-Cookie` 헤더가 붙지 않습니다.
~~~
$ tail -1 files/config/db.config.php
$db_info->use_session_conditionally=true;
....
$ curl -I  http://localhost/
HTTP/1.1 200 OK
Content-Type: text/html; charset=UTF-8
X-Powered-By: PHP/5.5.9-1ubuntu4.9
Cache-Control: public, must-revalidate, post-check=0, pre-check=0
X-Backend-Status: normal
Date: Thu, 09 Jul 2015 00:59:14 GMT
X-Varnish: 1471735403 1471735392
Age: 6
Via: 1.1 varnish
Connection: keep-alive
X-Backend-Server: nginx/1.4.x (Ubuntu)
Server: nginx/1.4.x (Ubuntu)
X-Cache-Hit: HIT
~~~
(몇차례 `curl -I`를 실행하면 `Cache-Hit`가 `HIT` 상태임을 나타내고 있습니다.)

ab 벤치마크 측정
~~~
$ ab -n 1000 -c 3  http://localhost/
This is ApacheBench, Version 2.3 <$Revision: 1528965 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking localhost (be patient)
Completed 100 requests
Completed 200 requests
Completed 300 requests
....
Requests per second:    8018.35 [#/sec] (mean)
Time per request:       0.374 [ms] (mean)
Time per request:       0.125 [ms] (mean, across all concurrent requests)
Transfer rate:          269428.95 [Kbytes/sec] received
...
~~~
8000 RPS로 무려 100배가 향상되었습니다.

동시에 200개 요청을 하고 20000회 요청을 해면 다음과 같습니다.
~~~
$ ab -n 20000 -c 200  http://localhost/
...
Requests per second:    14025.98 [#/sec] (mean)
Time per request:       14.259 [ms] (mean)
Time per request:       0.071 [ms] (mean, across all concurrent requests)
Transfer rate:          473883.72 [Kbytes/sec] received
~~~
무려 14000 RPS나 나옵니다. 초당  470MB/s
동시에 500개 요청을 해도 결과는 비슷합니다. (ab -n 20000 -c 500)

더 놀라운 것은 시스템의 부하 상태입니다. (ab -n 20000 -c 500을 처리하고 난 이후)
~~~
$ w
 10:04:25 up xx days, 17:28,  x user,  load average: 0.25, 0.19, 0.18
~~~

그러면 nginx만의 성능은 어떨까요 ? `ab -n 10000 -c 50 http://localhost:8000/` 1만회 50개 요청을 동시에 해보았습니다.

~~~
$ ab -n 10000 -c 50  http://localhost:8000/ (ngix가 8000 포트로 설정되어 있음)
This is ApacheBench, Version 2.3 <$Revision: 1528965 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking localhost (be patient)
Completed 1000 requests
Completed 2000 requests
Completed 3000 requests
Completed 4000 requests
....
Requests per second:    106.64 [#/sec] (mean)
Time per request:       468.850 [ms] (mean)
Time per request:       9.377 [ms] (mean, across all concurrent requests)
Transfer rate:          3570.91 [Kbytes/sec] received
....

~~~
100RPS 수준이고 3MB/s 밖에 안나옵니다.

그리고 시스템 부하상태입니다.
~~~
$ w
 10:09:01 up xx days, 17:33,  x user,  load average: 3.64, 1.45, 0.63
~~~

kijin commented:
XE는 프레임워크입니다. 게시판 등 흔히 사용하는 기능들이 기본으로 포함되어 있긴 하지만, 그것보다도 더 중요한 목적은 서드파티 개발자들이 자유롭고 편리하게 모듈과 레이아웃 등을 개발할 수 있도록 멍석을 깔아주는 것입니다. 세션을 자동으로 시작하고 로그인 상태를 알아서 관리해 주는 이유도 그거고요.

예를 들어 며칠 전에 통합검색에 관련된 XSS 취약점이 발견되었습니다. 1.8.4 버전에서는 레이아웃의 특정 부분에 필터링 코드를 추가하는 방법으로 막았지만, 이렇게 하면 모든 서드파티 레이아웃을 다 수정해야 하는 문제가 생깁니다. 그래서 1.8.5 버전에서는 코어 자체에서 특정 변수를 무조건 필터링해 주도록 하여, 기존의 레이아웃을 모두 그대로 사용할 수 있도록 변경되었습니다.

단 한 줄이라도 기존의 서드파티 자료들을 수정할 것을 요구해서는 안 되기 때문입니다.

예전에는 이런 원칙을 잘 지키지 않아서 1.4 → 1.5 버전업 당시 엄청나게 큰 문제가 발생했습니다. 이제 대세로 자리잡은 semantic versioning 원칙을 따른다면 서드파티 자료의 수정을 요구할 경우 메이저 버전의 변경이 필요합니다. 예전에 했던 것처럼 semantic versioning을 개무시하고 진행한다 해도 최소한 마이너 버전은 올려줘야 합니다. 1.8.x에서 해결할 수 없는 문제라는 거죠.

XE 코어의 개발에 참여하시는 분들이 빠른 어플리케이션을 만드는 법을 몰라서 이렇게 느려터지고 복잡하기 짝이 없는 코드베이스를 계속 유지보수하고 있는 걸까요? 그건 아니죠. **이런 것들은 코어에서 책임지고 처리해 주겠다, 이런 함수와 이런 클래스들을 반드시 제공하겠다고 7~8년 전에 약속했기 때문에** 그 약속을 지키기 위해 그대로 유지하고 있는 것입니다. 아무리 `ab` 벤치마킹 결과가 백배 천배 좋아진다고 해도, 세션 시작을 각 모듈에 떠넘기는 것은 이 약속을 저버리는 일입니다.

따라서 님의 패치는 XE 1.x보다는 다음 메이저 버전에 적용하는 것이 더 적합하다고 생각됩니다. 다음 메이저 버전은 1.x 버전의 약속을 지킬 필요가 없기 때문이지요. 정말로 체감성능을 100배 향상시킬 수 있다면 XE 1.x 사용자들을 다음 메이저 버전으로 끌어들이는 데도 좋은 당근이 될 수 있을 것 같아요.

비판은 여기까지 하고, 좀더 긍정적인 면에서 검토해 보자면...

1. Cache-Control 헤더를 일괄 적용하지 않고 각 뷰에서 선택할 수 있도록 하는 것은 아주 훌륭한 기능입니다. 기존 자료와의 호환성을 위해 기본값은 여전히 기존 방식을 따라야겠지만, 기본값이 아닌 Cache-Control 헤더를 선택할 수 있도록 옵션을 제공하는 것은 찬성합니다. 특히 글에 첨부된 이미지를 다운로드할 때는 Cache-Control, ETag, Last-Modified 등의 헤더를 잘 활용하면 불필요한 반복 다운로드를 막는 데 큰 도움이 되겠습니다.
2. 로그인하지 않은 사용자의 세션 시작 여부를 각 모듈의 판단에 맡기지 말고, 코어에서 좀더 지능적으로 처리해 주는 방법은 없을까요? 예를 들어 `Context::close()` 시점에서 아직 세션이 시작되지 않았는데도 `$_SESSION` 초전역변수가 존재한다면, 모듈에서 세션에 뭔가를 저장하려고 시도했다는 뜻이므로 코어에서 자동으로 세션을 시작하고 그 값을 대신 넣어주는 방법이 있겠습니다. 이렇게 구현한다면 서드파티 자료의 수정을 요구할 필요가 없습니다.
3. 코어 수정 없이 `Context` 클래스 실행 과정을 커스터마이징하기가 너무 어렵다는 점은 저도 통감하고 있습니다. 세션 쉴드, 언어 자동 선택 애드온 등을 개발하다 보니 불편하기 짝이 없더라구요. 기존 자료와의 호환성을 깨뜨리지 않으면서도 애드온이 끼어들 수 있는 시점을 좀더 제공한다면 굳이 코어를 수정하지 않더라도 님이 원하시는 것과 같은 기능을 애드온으로 충분히 구현할 수 있겠습니다.

YJSoft commented:
@kijin 문제는 `다음 메이저 버전`이 없다는 점입니다.
XE 2.0은 취소된 버전이며 XE 3은 개발중입니다. 1.9로 릴리즈하기에도 변경사항이 너무 큽니다.
즉, 이 수정사항에 매길 수 있는 버전이 없습니다.

개인적으로 이제 XE는 안정화 단계를 넘어서 유지보수 단계로 들어왔다 생각하고 있습니다. 즉, 더이상의 큰 변경사항은 없어야 한다는 것이죠.

kijin commented:
@YJSoft 제가 얘기한 `다음 메이저 버전`이 바로 XE3입니다 ^^

XE3에서는 캐싱을 감안한 세션, 쿠키, 헤더 처리가 이루어졌으면 좋겠다. 그러나 1.8에 이 패치를 적용해 달라고 요구하는 것은 적절하지 않다... 라는 의미였습니다.

YJSoft commented:
@kijin XE3은 라라벨 기반이니, 적어도 캐싱 관련 내용을 하드코딩해버린 XE보다는 낫지 않을까요?
@wkpark 당장 급한 문제는 아니니 조금 여러가지로 고민해 보고 결정해야 할 것 같습니다. 일단 코딩 컨벤션 문제도 있고요.

wkpark commented:
XE3 라라바는 이미 varnish를 잘 지원 합니다. 네. 프레임웨크 레벨에서 잘 지원 하겠지요. 세부적으로 각 모듈 구현이 잘못되어서 `$_SESSION`을 남발한다면 vanish의 성능이 형편없이 떨어질 수밖에 없는 것이고요.

XE에서 이를 지원할 계획이 없다면 당장에 이 패치가 필요한 사용자들을 위해 포크를 할 수밖에는 없을듯합니다. 패치 수준의 관리가 되겠지만요.
아직 제가 xe-core 소스를 정확히 이해하지 못한 상황에서 좀 더 좋은 해결법에 대한 토론 없이 이렇게 논쟁으로 그치고 계속 지지부진하다면  차라리 좀 더 좋은 방식이 나오기 전까지 따로 소스를 유지하는 것이 낫겠습니다.
이 방식은 너무나 간단하며, xe-core를 좀 더 잘 이해한다면 하위 호환 문제가 없는 현명한 방식이 충분히 가능하리라 생각합니다.

이 방식으로 varnish 서버와 결합해서 쓰면 당장에 동접 500도 쩔쩔맬 상황에서 500동접은 식은죽먹기가 되고 서버관리비도 덜나가게 되는데 그러한 현실적인 문제를 생각하지 않을 수가 없습니다.

ForPeople commented:
저는 이 논쟁이 잘 이해가 안되는게 옵션 처리 해주고 있는 기능인데 왜 기존의 모듈 개발자를 고려해서 기능 추가를 하면 안된다고 하시는지 잘 모르겠습니다. 이게 버전 1.9로 가야 할 만한 사안인가요? 옵션을 안키면 되지 않나요?

wkpark commented:
@ForPeople // 네 제가 주장하고 있는 내용입니다. 옵션으로 처리하고 장기적으로 고치면 되고, 고치는것도 쉽다는 말이 제 요점입니다.

qw5414 commented:
@ForPeople @wkpark 
```
XEUser1 : (생각 - 오 이거 채크하면 빨라진다네. 그다음 내용은 뭔소리야 몰라 넘어가). 어라.. 에러다 업데이트 했더니 서드 파티에서 오류있네 xe1.8.5 뭐임?
XEUser2 : 이거 뭐지. 일단 해봐야지.(옵션건듬.) 어라 출석부랑 다른 모듈에서 에러 나네.. 데이터도 안오고.. 뭐때문이지?
```

이와 같은 상황이 연출되니깐요. 단순 옵션을 경고해도 속도관련에서는 한국사람들은 민감하기 때문에 여러가지 만져보게 되고 그중에 옵션을 사용할 가능성이 큰데, 그 이후 에러의 원인을 초보자는 모른다는겁니다.
그리고, 옵션을 진행하는데 있어서 애초에 서드파티의 모듈이 수정이 되어야할 만큼의 문제가 있는것인데, 이 부분을 옵션을 지원한다 안한다를 떠난 논외문제점이라 판단되는데요..

kijin commented:
@ForPeople `$db_info`에 들어가는 옵션이라면 관리모듈 → 설정 → 기본 → 고급 섹션에 "인증 세션 DB 사용" 바로 밑에쯤 들어갈 법한 옵션인데, 현재 여기에 들어가는 옵션들을 사용자가 어떻게 바꾸더라도 서드파티 모듈의 정상 작동에는 영향을 미치지 않습니다. 영향을 미치지 않아야 하는 것이 당연한 일이고요. "짧은 주소 사용 옵션을 켜면 안 되는 모듈", "SSL 항상 사용을 선택하면 안 되는 모듈", "varnish 캐싱 옵션을 켜면 안 되는 모듈" 등을 사용자가 일일이 신경써 가며 설치해야 하겠습니까? 최소 버전만 충족하면 다 되어야 정상이죠.

@wkpark 서드파티 모듈을 전혀 수정하지 않고도 코어 내에서 구현하는 방법이 있을 것 같다고 위에서 말씀드렸습니다. 생각해 보셨는지요? 다른 사람에게 한 줄 고치라고 말하기는 쉽습니다. 그러나 100명이 한 줄씩 고칠 필요가 없도록 하기 위해 내가 100줄을 고치는 귀찮음을 무릅써야 하는 것이 프레임워크의 운명입니다. 게다가 세션은 프레임워크가 처리해 주기로 약속한 부분의 기본 중에서도 가장 기본이 되는 기능입니다. 세션 시작을 각 모듈에 떠넘기는 것은 심각한 책임전가입니다.

qw5414 commented:
저한테 맨션으로 보냈던 내용중에서도 마찬가지입니다.
XE는 우리같이 코드를 수정할 수 잇고 어떤 구조상으로 돌아가는지 아는 사람들이 많은 반면에, 아무것도 모르는 사람이 대부분이라는 겁니다.

당장 XE묻고 답하기 게시판에 글이 엉청나게 게시되는것만 봐도요..

당장 세션을 사용하는 모듈이 뭐가 있는지 재대로 아는 사람 몇명이나 될까요?
개발자도 알까요? 코드를 안본다면 자기가 사용하는 모듈이 자기가 XE개발자라 하더라도 모를테지요.

그러한 사항에서 세션에 문제있으므로 쓸사람은 각모듈의 세션을 시작하는 코드를 추가 해야하는거에요.

그것도 자신이 쓴것마다요...

그걸 일일이 다 찾고 수정하는건 저도 솔찍히 좋은 생각으로 받아들여지지 않네요.

단순 속도를 위한패치를 하기위해 그 노력을 한다는건 좀 저도 이상하다는 생각들거든요..

devwit commented:
단일 사이트의 개인서버라면 도전해볼 하겠네요

ForPeople commented:
@wkpark

저는 개인적으로 와이낫? 하는 입장이지만, 다른 분들의 의견을 종합해서 타협해 보자면 별도의 모듈을 이용해서 켜고 끄는 기능을 만들어 구현이 되어야 할 것 같습니다. 컨텍스트부분에서 세션을 켜고 끄는 것이 애드온 부분에서 호출 가능한가요??? 모듈 + 애드온 등으로 구현이 가능하다면 그리 하시고 공홈에 별도의 자료로 배포하시는게 좋을 것 같습니다.

wkpark commented:
사실 이 문제를 근본적으로 해결하려면 `$_SESSION`을 direct로 엑세스하면 안되고 모듈로 처리하는 방식을 쓰면 깔끔합니다. XE도 기본적으로 프레임워크이고 이미 세션 클래스가 준비되어있습니다. 그런데 이것에 맨처음에 도입된게 아니고, 2008년에 세션 클래스가 도입이 되었고, `$_SESSION`을 쓰는 방식이 아닌 DB를 쓰는 방식이더군요. 그러나  이를 제대로 쓰지 않고 있으며, `$_SESSION`을 직접 엑세스하는 부분이 xe-core까지 여전히 남아있습니다. (250여군데) (`select * from xe_session;` 하면 결과값이 없음) 코어에 이정도까지 남아있으니 애드온에서 `$_SESSION` 남발한다고 뭐라 할 사안이 아니긴 합니다.

라라벨같은 경우는 `$_SESSION`을 엑세스하지 않으며 `get()`, `set()`으로 키-밸류를 다루는 클래스를 씁니다. 라라벨 소스를 뒤져보면 놀랍게 `$_SESSION`이 없습니다. https://github.com/symphonycms/symphony-2/blob/master/symphony/lib/core/class.session.php 심포니의 세션 클래스를 보면 `session_id() == ''`일 경우에 선택적으로 `session_start()`가 호출되는 것을 볼 수 있습니다.

코어에서 고칠 수 있는 깔끔한 방식이 없을지 더 고민해보도록 하겠습니다.

wkpark committed:
checkSessionUsed() 메쏘드 추가하여 $_SESSION이 사용된 경우 session_start() 자동으로 실행되게 함

wkpark commented:
#1613 이슈에서 하위호환 보장하는 방식을 제안하였습니다.

kijin commented:
#1615 이슈에서 **진짜로** 하위호환 보장하는 방식을 제안하였습니다 ^^

wkpark commented:
XE + Varnish 캐시서버를 사용하실 분들을 위해서 간단한 설명서를 작성했습니다.

https://github.com/wkpark/xe-core/wiki/XeWithVarnishCache

wkpark commented:
패치 모두 걷어내고, 애드온의 경우에 `$_SESSION`을 사용하는 경우 자동으로 `session_start()`가 호출되도록 하였습니다. 따라서 session_start()를 애드온에서 신경안써도 됩니다.

나머지 코어에서 `$_SESSION`을 세심하게 제어하지 않는 부분은 버그로 볼 수 있을것이며, 이 부분에 대해서는 #1613 및 #1615 등을 참고해서 고쳐야 합니다.

wkpark committed:
openSession(), statSession(), getSessionStatus() method added

wkpark committed:
$_SESSION을 필요한 경우에만 세팅

wkpark committed:
lazyStartSession() 추가. 세팅된 $_SESSION을 검사하여 세션을 자동으로 시작함

wkpark committed:
$_SESSION['lang_type'] 삭제: commit 90aa62a6acbce5725 "위젯 검출 정규 표현식을 수정"에서 추가되었다가 현재 더이상 안쓰임

wkpark committed:
_checkAccessibleFromStatus() 직접 호출하고 $_SESSION사용 배제함

wkpark commented:
여기 저기 소스를 볼수록 이걸 어떻게 고쳐야 할지 감이 잡히네요~

당분간 이 패치는 묶혀두고 다른 부분을 고쳐야겠습니다

wkpark commented:
nginx reverse cache proxy + XE + 아파치 세팅을 해도 거의 비슷한 효과를 볼 수 있네요. 단, nginx의 경우에는 설정을 조금 조절해야 합니다. varnish는 설정을 기본값을 그대로 써도 효과를 볼  수 있습니다.

~~~
                proxy_no_cache $cookie_nocache $arg_nocache$arg_comment; # 쿠키쓰면 캐싱 안함
                proxy_cache_valid 200 3s; # 3초간 캐싱. (이 값이 너무 크면 컨텐츠가 제때 업뎃이 안됨)
~~~

ioz2014 commented:
@qw5414
포인트 모듈의 사용을 옵션화한 것도 그렇게 따지면 불가능하지 않을까요? 포인트 트리거에 의존하는 모듈 중 가령 포인트로 상품을 사는 아이콘샵, 아이템샵(이건 안써봐서 잘 모르지만)은 당연하게도 정상 작동이 안되겠죠. XE에 대해 잘 모르는 분들에게는 똑같이 '뭐야 이거 왜 되던게 안돼!!' 라는 반응이 나올 수 있다고 봅니다.
이런걸 막기 위해서는 키고 끄는 옵션 부분에 설명을 넣어야겠죠. [잘 모르시겠으면 사용하지 마세요!] 라고 볼드체로 써놔도 될거고...

@kijin 포피플님께 남기신 멘션에서 옵션을 꺼도 영향을 받는 서드파티 모듈이 없다고 하셨는데 가령 예를 드신 인증세션 DB만 봐도 이걸 끄고 키는데에 따라서 세션DB에 의존하는 접속자수 모듈이나 위젯이 영향을 받겠죠..

그런데... 방문자에게 매기던 세션까지 사라지면 당장 게시판 조회수가 문제가 되겠군요. 조회기록을 저장하지 않는 이상 새로고침할때마다 올라갈테니... ㅋ 쿠키로 대체할 수는 있지만 그러려면 board 모듈 업데이트가 필요하겠네요. 뭐... 이런 식으로.. 서드파티 모듈의 수정이 크게 발생할 수 있다고는 생각합니다. 기진곰님이 제안하신 하위호환으로도 근본적인 문제는 해결할 수 없겠죠.

wkpark commented:
현재 하위호환문제는 모두 해결되었습니다.

옵션도 아예 필요하지 않게 재 수정되었습니다.


세션을 안쓰자는게 아니예요.
캐싱을 방해하는 세션콜을 최소화 하는 고침이 이 패치입니다.

kijin commented:
@ioz2014 포인트 모듈이나 게시판 모듈은 서드파티(third-party)가 아닙니다. 코어와 함께 관리·배포되는 퍼스트파티(first-party) 모듈이므로, 코어에 새로운 기능이 추가될 경우 거기에 맞춰서 수정하여 배포할 수 있습니다. 호환성 문제가 발생하는 것은 코어의 변경내역을 일일이 따라가기 곤란한 서드파티 모듈 뿐입니다.

- 제1자(퍼스트파티) : XE 코어 개발팀
- 제2자(세컨트파티) : 최종 사용자
- 제3자(서드파티) : 그 밖에 XE와 연동되는 모듈, 애드온 등을 개발하는 사람을 의미합니다.

> 영향을 받겠죠..

옵션이라는 것은 XE의 작동 방식에 의도적인 변화를 가하는 것이므로, 일부 모듈의 작동 방식에 영향을 준다고 모두 나쁜 것은 아닙니다. 영향을 주지 않는다면 애초에 옵션을 제공하는 의미가 없죠.

예를 들어 인증세션 DB 사용 옵션을 사용하지 않을 경우에도 정확한 접속자 수 집계가 어려운 것은 마찬가지입니다. 그렇다고 그 옵션이 나쁘다고 하지는 않잖아요. 그 옵션 사용 여부에 따라 방문 횟수나 조회수 등이 정확하게 집계되지 않을 수도 있다는 안내문이 필요할 뿐입니다.

어떤 옵션이 다른 프로그램에 미치는 영향에는 2가지가 있습니다.

1. 그 옵션의 원래 의도
2. 의도하지 않은 부작용

접속속도가 빨라지는 것은 원래 의도입니다. 조회수가 정확하게 집계되지 않는 것도 원래 의도에 포함됩니다. 애초에 캐시서버를 사용한다는 것은 XE단에서 조회수를 정확하게 집계할 수 없게 만든다는 의미이니까요. 이런 것들은 문제가 되지 않습니다.

만약 포인트마저 정확하게 집계되지 않는다거나, 서드파티 쇼핑몰 모듈에서 장바구니에 물건을 담을 수 없다거나, 이렇게 된다면 부작용에 해당하겠지요. 이런 일이 있어서는 안 됩니다.

포인트 모듈의 경우, 로그인하지 않은 사용자에게도 포인트를 부여하고 그 포인트를 사용할 수 있도록 해주는 기능이 있었던가요? 포인트라는 것은 대개 회원가입을 하고 로그인한 상태에서만 사용할 수 있는 것이 아닌가요? 그렇다면 로그인시 세션을 시작해 주는 것으로 충분합니다.

> 세션을 필요로하는 모듈 하나를 불러오기만 해도 근본적인 이 패치의 취지는 사라지지 않을까요?

당연히 그래야죠. 어떤 모듈을 사용하느냐 하는 것은 각 사이트 운영자의 고유권한이니, 캐시서버를 사용해서 속도를 대폭 향상시키고 싶은 운영자라면 세션을 필요로 하는 모듈 사용을 피하겠지요. 신경쓰지 않는 운영자라면 그냥 다 써도 되고요. 모듈에 따라 성능에 큰 차이가 있는 것은 지금도 마찬가지이므로, 딱히 달라지는 것도 없을 듯 합니다.

wkpark commented:
이 패치는 이미 처음 제안했던 방식의 문제를 다시 고쳐서 하위 호환성 문제가 전혀 없게 고쳐졌습니다
이와 아울러 비슷한 방식의 다른 접근을 #1613 및 #1615 에서 하고있습니다.

기본적으로 이 방식이 하는 것은 `session_start()`를 무조건 호출하는 현 방식을 개선해서 `session_start()`가 선택적으로 호출되게 하는 것입니다.

@kijin님이 이것의 효과에 대해서 너무 잘 설명해주셨기 때문에 저는 세션의 작동방식에 대해서 좀 설명하도록 하겠습니다.

* 세션을 시작하려면 `session_start()`를 호출해야 합니다. `session_start()`이 호출되면 세션이 비로소 시작이 되며 `$_SESSION` 쿠키를 접근할 수 있게 됩니다.
* 일반적인 상황에서는 `session_start()`의 호출이 불필요합니다.
* 세션이 시작되지 않은 경우에는 `$_SESSION` 변수 자체가 비어있으므로 이 상태에서 `session_start()`를 호출해도 `$_SESSION`쿠키가 텅 비어있는 상태입니다.
* 세션이 사작된 이후에 `$_SESSION` 쿠키에 값을 세팅하면 이 값은 자동으로 별도의 저장소에 저장이 됩니다. (PHP는 기본적으로 이 값을 파일로 저장)
* 세션이 시작된 이후에는 `$_COOKIE['PHPSESSID']`값이 자동으로 세팅이 됩니다. 이 값이 세팅이 되어있는 경우에는 `session_start()`를 항상 호출해야 `$_SESSION` 쿠키를 접근할 수 있습니다.
* `$_COOKIE['PHPSESSID']가 비어있다면 `session_start()`를 호출하지 않아도 됩니다.
* 문제는 `session_start()`가 호출되지 않았는데 `$_SESSION`에 값을 저장하려는 경우입니다. 이 경우에 `session_start()`가 먼저 호출되어야 `$_SESSION`에 값이 저장되며, 그렇지 않은 경우에는 `$_SESSION`쿠키에 세팅한 값이 날아가 버리고, 오류가 나게 됩니다.
* 그래서 이 문제를 해결하기 위해서 `lazyStartSession()`을 새롭게 추가했습니다. `lazyStartSession()`은이미 세션이 시작된 경우에는 무시하고, 세션이 시작되지 않았는데 `$_SESSION`쿠키가 세팅이 되었다면, `session_start();`를 비로소 호출하고 세팅된 `$_SESSION`값을 재설정합니다. 이렇게 하므로서 하위 호환성을 확보하였습니다.
* 한번 `session_start()`가 호출이 되고나면 `$_COOKIE['PHPSESSID']`가 자동으로 세팅되며, `session_start()`는 자동으로 호출되게 됩니다.

이것이 세션의 작동 원리입니다. 이것에 의하면 `session_start()`가 무분별하게 호출될 필요가 없고, `$_COOKIE['PHPSESSID']`를 검사해서 선택적으로 `session_start()`를 호출하면 되고, 또한 `lazyStartSession()`에 의해서 하위 호환문제를 제거하였습니다.

또한 애드온에서 `$_SESSION`을 호출하고 바로 종료해버리는 케이스(`lazyStartSession()`이 제대로 호출되지 않는 케이스)를 대비하기 위해서, 애드온을 자동 로딩하는 부분을 수정하여 애드온에서 `$_SESSION`을 사용하는 경우를 자동으로 검사하여 해당 애드온에서 무조건적으로 `lazyStartSession()`이 호출되도록 하여 하위 호환 문제를 제거하였습니다.

wkpark committed:
Context::setCacheControl()를 DisplayHandler로 옮김

 * setCacheControl()를 DisplayHandler에서 호출

wkpark committed:
partial revert

wkpark committed:
Revert "checkSessionUsed() 메쏘드 추가하여 $_SESSION이 사용된 경우 session_start() 자동으로 실행되게 함"

This reverts commit ebecfcdc4e1539b081bf340f35b7ffdd293c65da.

wkpark committed:
비어있는 $_SESSION 쿠키 검사하지 않고 복사함. 간소화

Please see also https://github.com/kijin/xe-core/commit/23a83a70330e6be66391037d4eb4eb40308ed69c

wkpark commented:
하위호환성 문제는 모두 제거되었으며, 이 패치는 최소한의 패치라고 보시면 됩니다. @kijin 님의 #1615 PR에는 여기에 몇가지 기능적인 부분이 더 추가된 부분만 빼놓고 이 패치와 대동소이합니다.

* 일반 비 로그인 사용자 - `session_start()` 자동 호출 안됨 - 캐시서버에서 캐시 가능해짐
* 세션을 사용하는 애드온을 쓸 경우 - 자동으로 판별하여 `session_start()` 선택적으로 호출됨

wkpark committed:
setCacheControl() 호출 위치 옮김
